soft4pes/control/mpc/solvers/mpc_enum.py

```diff
@@ -0,0 +1,171 @@
+"""Enumeration based solver for Model Predictive Control (MPC)."""
+
+from itertools import product
+import numpy as np
+
+
+class MpcEnum:
+    """
+    Enumeration based solver for Model Predictive Control (MPC).
+
+    Attributes
+    ----------
+    J : 1 x conv.nl^(3*Np) ndarray of floats
+        Array for costs.
+    u_seq : 3*Np x conv.nl^(3*Np) ndarray of ints
+        Array for 3-phase switch position sequences.
+    i : int
+        Counter for the switch position sequences.
+    SW_COMB : 1 x conv.nl^3 ndarray of ints
+        All possible 3-phase switch position.
+    """
+
+    def __init__(self, conv):
+        """
+        Initialize an MpcEnum instance.
+
+        Parameters
+        ----------
+        conv : converter object
+            Converter model.
+        """
+        self.J = None
+        self.u_seq = None
+        self.i = 0
+
+        if conv.nl == 2:
+            sw_pos_3ph = np.array([-1, 1])
+
+        elif conv.nl == 3:
+            sw_pos_3ph = np.array([-1, 0, 1])
+
+        # Create all possible 3-phase switch position
+        self.SW_COMB = np.array(list(product(sw_pos_3ph, repeat=3)))
+
+    def __call__(self, sys, conv, ctr, ref):
+        """
+        Solve MPC problem by using exhaustive enumeration.
+
+        Parameters
+        ----------
+        sys : system object
+            System model.
+        conv : converter object
+            Converter model.
+        ctr : controller object
+            Controller model.
+        ref : ndarray of floats
+            Reference.
+
+        Returns
+        -------
+        uk : 1 x 3 ndarray of ints
+            The 3-phase switch position sequence with the lowest cost.
+        """
+
+        # Initialize array for costs and 3-phase switch position sequences
+        self.J = np.zeros((conv.nl**(3 * ctr.Np), 1))
+        self.u_seq = np.zeros((conv.nl**(3 * ctr.Np), 3 * ctr.Np))
+        self.i = 0
+
+        self.solve(sys, conv, ctr, sys.x, ref, ctr.u_km1, 0)
+
+        # Find the 3-phase switch position sequence with the lowest cost
+        min_index = np.argmin(self.J)
+        uk = self.u_seq[min_index, 0:3]
+        return uk
+
+    def solve(self, sys, conv, ctr, xk, ref, u_km1, k):
+        """
+        Recursively compute the cost for different 3-phase switch position sequences.
+
+        Parameters
+        ----------
+        sys : system object
+            System model.
+        conv : converter object
+            Converter model.
+        xk : ndarray of floats
+            Current state vector.
+        ref : 1 x 2 ndarray of floats
+            Reference.
+        u_km1 : 1 x 3 ndarray of ints
+            Previous 3-phase switch position.
+        k : int
+            Prediction step.
+        """
+
+        # Calculate the range of columns for the current prediction step
+        # This is done in order to save calculation effort, as redundant calculations are avoided
+        u_col_range_start = 3 * k
```

Comment:
If we look at the possible switching sequences for a case Np=2, we see that for the 8 first sequences (orange), the same 3-phase switch position [-1 -1 -1] repeats. It does not make sense to calculate the same result 8 times, which is why the redundancy is utilized. The same repeats for the green positions and so on. This extends to higher prediction horizons as well.

Of course, if we wish, for clarity and example's sake, the utilization of redundancy can be removed and the solver will calculate every sequence from start to finish.

![image](https://github.com/TAU-Power-Electronic-Systems/Soft4PES/assets/97534513/b599baff-d44c-4e5a-8aff-7a3fd99991a0)

Reply:
I see your point now that you described your implementation. I haven't debugged so I am not sure how it works. However, would it be simpler if we just had a recursive implementation? Then, assuming N_p = 2, you check one three-phase witch position in the first step of the horizon (e.g., [-1 -1 -1]^T), and then compute the state-update and intermediate cost (stage cost for the first step). Following you go to the next horizon step where you check all possibilities one by one (8 new stage costs added to the already stored stage cost from the first step with [-1 -1 -1]^T). From this, you an compute a first final cost for the complete objective function. Then you go one level up (again first step of the horizon), assign the next option (e.g., [-1 -1 1]^T), compute the new stage cost, etc. This way you can easily do some basic branch-and-bound as well.

Reply:
The implementation is actually how you described it. The enumeration is recursive.

```
        u_col_range_start = 3 * k
        u_col_range_end = 3 * (k + 1)
        u_col_range = range(u_col_range_start, u_col_range_end)
        u_row_range = range(self.i, self.i + conv.nl**(3 * (ctr.Np - k - 1)))
```
These rages are used to fill the u_seq matrix with the 3-phase switch positions and the J vector with costs. The figure in the previous message represents that. On the first prediction step and  u = [-1 -1 -1]^T, k = 0 and i = 1. With these, we fill the orange part of the u_seq matrix and the corresponding rows of J vector

Of course, it would be possible to save only the lowest cost and the switch position sequence, but that would be quite close to the branch and bound algorithm

Reply:
Ok, I see your point. It is hard to debug in this form, but when the whole simulation environment is ready hopefully thing will be easier. II think it would be nice to build some very basic branch-and-bound method as I described in my previous comment